File: LeetCode/134GasStation/GasStation.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        // Greedy -> Time:O(n), Space:O(1)
        int total = 0, start = 0, gas_sum = 0, cost_sum = 0, n = gas.size();
        for (int i = 0; i < n; i++) {
            gas_sum += gas[i];
            cost_sum += cost[i];
        }
        if (gas_sum < cost_sum) {
            return -1;
        }
        for (int i = 0; i < n; i++) {
            total += gas[i] - cost[i];
            if (total < 0) {
                total = 0;
                start = i + 1;
            }
        }
        return start;
    }
};

File: LeetCode/134GasStation/GasStation.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        # Greedy solution -> Time:O(n), Space:O(1)
        if sum(gas) >= sum(cost):
            start = 0
            diff = [gas[i] - cost[i] for i in range(len(gas))]
            total = 0
            for i in range(len(gas)):
                total += gas[i] - cost[i]
                if total < 0:
                    total = 0
                    start = i + 1
            return start
        return -1

        # # Brute force -> Time:O(n^2), Space:O(1)
        # for i in range(len(gas)):
        #     if gas[i] > cost[i]:
        #         current_station = i
        #         tank = gas[current_station]
        #         while current_station != i - 1:
        #             tank -= cost[current_station]
        #             # travel to next station
        #             current_station = (current_station+1)%len(gas)
        #             tank += gas[current_station]
        #         if tank >= cost[current_station]:
        #             return i
        # return -1
